internal/service/resources/collector/collector.go

```diff
@@ -0,0 +1,265 @@
+func (c *Collector) Run(ctx context.Context) error {
+	if err := c.init(ctx); err != nil {
+		return err
+	}
+
+	// Run the initial data collection
+	c.execute(ctx)
+
+	for {
+		select {
+		// TODO: Add hook for new data sources from watch events
+		// TODO: Add hook for kick to run collection based on watch events on individual data sources
+		case <-time.After(pollingDelay):
+			c.execute(ctx)
+		case <-ctx.Done():
+			slog.Info("Context terminated; collector exiting")
+			return nil
+		}
+	}
+}
+
+// init runs the onetime initialization steps for the collector
+func (c *Collector) init(ctx context.Context) error {
+	slog.Info("initializing data collector")
+
+	for _, d := range c.dataSources {
+		err := c.initDataSource(ctx, d)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// initDataSource initializes a single data source from persistent storage.  This recovers its unique UUID and
+// generation ID so that it continues from its last save point.
+func (c *Collector) initDataSource(ctx context.Context, dataSource DataSource) error {
+	name := dataSource.Name()
+	record, err := c.repository.GetDataSourceByName(ctx, name)
+	switch {
+	case errors.Is(err, utils.ErrNotFound):
+		// Doesn't exist so create it now.
+		result, err := c.repository.CreateDataSource(ctx, &models.DataSource{
+			Name:         name,
+			GenerationID: dataSource.GetGenerationID(),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create new data source %q: %w", name, err)
+		}
+
+		dataSource.SetID(*result.DataSourceID)
+		slog.Info("created new data source", "name", name, "uuid", *result.DataSourceID)
+	case err != nil:
+		return fmt.Errorf("failed to get data source %q: %w", name, err)
+	default:
+		dataSource.SetID(*record.DataSourceID)
+		dataSource.SetGenerationID(record.GenerationID)
+		slog.Info("restored data source",
+			"name", name, "uuid", record.DataSourceID, "generation", record.GenerationID)
+	}
+	return nil
+}
+
+// execute runs a single iteration of the main loop.  It does not return an error because all errors should be handled
+// gracefully.  If a truly unrecoverable error happens then a panic should be used to restart the process.
+func (c *Collector) execute(ctx context.Context) {
+	slog.Debug("collector loop running", "sources", len(c.dataSources))
+	for _, d := range c.dataSources {
+		slog.Debug("collecting data from data source", "source", d)
+		if err := c.executeOneDataSource(ctx, d); err != nil {
+			slog.Warn("failed to collect data from data source", "source", d, "error", err)
+		} else {
+			slog.Debug("collected data from data source", "source", d)
+		}
+	}
+	slog.Debug("collector loop complete", "sources", len(c.dataSources))
+}
+
+// executeOneDataSource runs a single iteration of the main loop for a specific data source instance.
+func (c *Collector) executeOneDataSource(ctx context.Context, dataSource DataSource) (err error) {
+	// Increment the generation id
```

Comment:
I think we are missing debug logs here at the start of  executeOneDataSource with the new generation ID and inside of collectResourcePools, collectResources and collectDeploymentManagers

Reply:
added